File: Tree/379-longestPath.py
class TreeNode:
    def __init__(self,val,left=None,right=None) -> None:
        self.val=val
        self.left=left
        self.right=right



ans=0     
def dfs(root):
    if not root:
        return 0
    
    l=dfs(root.left)
    r=dfs(root.right)
    # nonlocal ans
    ans=max(ans,1+r+l)
    return max(l,r)+1



def longestPath(root):
    if not root:
        return 0
    
    dfs(root)
    # nonlocal ans
    return ans


one=TreeNode(1)
two=TreeNode(2)
three=TreeNode(3)
four=TreeNode(4)
five=TreeNode(5)
six=TreeNode(6)
nine=TreeNode(9) 


one.left=two
one.right=three

two.left=six
three.left=4
three.right=9
four.right=5

print(longestPath(one))   

File: Tree/379-longestPath.cpp
#include"../treeNode.h"
#include<iostream>


int foo(TreeNode* root,int &ans)
{
    if(!root)
        return 0;
    
    int l=foo(root->left,ans);
    int r=foo(root->right,ans);
    ans=std::max(ans,1+l+r);
    return std::max(l,r)+1;
}

int longestPath(TreeNode* root)
{
    if(!root)
        return 0;
    int ans=0;
    foo(root,ans);
    return ans;
}

int main()
{
    TreeNode *one=new TreeNode(1);
TreeNode* two=new TreeNode(2);
TreeNode* three=new TreeNode(3);
TreeNode* four=new TreeNode(4);
TreeNode *five=new TreeNode(5);
TreeNode *six=new TreeNode(6);
TreeNode *nine=new TreeNode(9); 


one->left=two;
one->right=three;

two->left=six;
three->left=four;
three->right=nine;
four->right=five;

std::cout<<longestPath(one);
}